### PR TITLE
[flink] Flink LookupTable support cache row filter

### PR DESCRIPTION
**Pull request by Sxnan**
### Purpose

Flink LookupTable support cache row filter to filter out rows to be stored in cache.

### Tests

- UT added in `LookupTableTest` and `KeyValueFileReadWriteTest`.

### API and Format

No

### Documentation

No

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/io/KeyValueFileReaderFactory.java [hunk 1]
```diff
@@ -102,11 +103,23 @@ public RecordReader<KeyValue> createRecordReader(DataFileMeta file) throws IOExc
 
     public RecordReader<KeyValue> createRecordReader(
             long schemaId, String fileName, long fileSize, int level) throws IOException {
+        return createRecordReader(schemaId, fileName, fileSize, level, null);
+    }
+
+    public RecordReader<KeyValue> createRecordReader(
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/io/KeyValueFileReaderFactory.java:109]: I think you don't need to modify `paimon-core`, you can do the filter outer.

Comment (Sxnan) [reply paimon-core/src/main/java/org/apache/paimon/io/KeyValueFileReaderFactory.java]: Good point! I updated the PR so we don't apply the filter in the `KeyValueFileReaderFactory`.  

However, I think we still need to modify the `LocalTableQuery` in the `paimon-core` unless I am missing something.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/FileStoreLookupFunction.java [hunk 2]
```diff
@@ -361,4 +367,8 @@ protected Set<Integer> getRequireCachedBucketIds() {
         // TODO: Implement the method when Flink support bucket shuffle for lookup join.
         return null;
     }
+
+    protected void setCacheRowFilter(@Nullable Filter<InternalRow> cacheRowFilter) {
```
Comment (wwj6591812) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/FileStoreLookupFunction.java:371]: Where this method be called?

Comment (wwj6591812) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/FileStoreLookupFunction.java:371]: What business scenarios is this feature used in?

Comment (Sxnan) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/FileStoreLookupFunction.java:371]: This change can reduce the memory usage of the lookup join function if how the fact table rows are shuffled is known by the lookup join function. We can filter out rows that will not shuffle to this specific subtask.

For this feature to work, Flink will support specifying a custom data distribution of the fact table, please see the [FLIP](https://cwiki.apache.org/confluence/display/FLINK/FLIP-462+Support+Custom+Data+Distribution+for+Input+Stream+of+Lookup+Join).

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**